src/Features/CSharp/Portable/MakeMethodAsynchronous/CSharpMakeMethodAsynchronousCodeFixProvider.cs

```diff
@@ -54,46 +55,93 @@ protected override bool IsAsyncSupportingFunctionSyntax(SyntaxNode node)
             return method.WithReturnType(newReturnType).WithModifiers(newModifiers);
         }
 
         private SyntaxNode FixLocalFunction(
             bool keepVoid, IMethodSymbol methodSymbol, LocalFunctionStatementSyntax localFunction,
-            INamedTypeSymbol taskType, INamedTypeSymbol taskOfTType, INamedTypeSymbol valueTaskOfTType)
+            KnownTypes knownTypes)
         {
-            var newReturnType = FixMethodReturnType(keepVoid, methodSymbol, localFunction.ReturnType, taskType, taskOfTType, valueTaskOfTType);
+            var newReturnType = FixMethodReturnType(keepVoid, methodSymbol, localFunction.ReturnType, knownTypes);
             var newModifiers = AddAsyncModifierWithCorrectedTrivia(localFunction.Modifiers, ref newReturnType);
             return localFunction.WithReturnType(newReturnType).WithModifiers(newModifiers);
         }
 
         private static TypeSyntax FixMethodReturnType(
-            bool keepVoid, IMethodSymbol methodSymbol, TypeSyntax returnType,
-            INamedTypeSymbol taskType, INamedTypeSymbol taskOfTType, INamedTypeSymbol valueTaskOfTType)
+            bool keepVoid, IMethodSymbol methodSymbol, TypeSyntax returnTypeSyntax,
+            KnownTypes knownTypes)
         {
-            var newReturnType = returnType.WithAdditionalAnnotations(Formatter.Annotation);
+            var newReturnType = returnTypeSyntax.WithAdditionalAnnotations(Formatter.Annotation);
 
             if (methodSymbol.ReturnsVoid)
             {
                 if (!keepVoid)
                 {
-                    newReturnType = taskType.GenerateTypeSyntax();
+                    newReturnType = knownTypes.TaskType.GenerateTypeSyntax();
                 }
             }
             else
             {
-                if (!IsTaskLike(methodSymbol.ReturnType, taskType, taskOfTType, valueTaskOfTType))
+                var returnType = methodSymbol.ReturnType;
+                if (IsIEnumerable(returnType, knownTypes) && IsIterator(methodSymbol))
+                {
+                    newReturnType = MakeGenericType("IAsyncEnumerable", methodSymbol.ReturnType);
+                }
+                else if (IsIEnumerator(returnType, knownTypes) && IsIterator(methodSymbol))
+                {
+                    newReturnType = MakeGenericType("IAsyncEnumerator", methodSymbol.ReturnType);
+                }
+                else if (IsIAsyncEnumerableOrEnumerator(returnType, knownTypes))
+                {
+                    // Leave the return type alone
+                }
+                else if (!IsTaskLike(returnType, knownTypes))
                 {
                     // If it's not already Task-like, then wrap the existing return type
                     // in Task<>.
-                    newReturnType = taskOfTType.Construct(methodSymbol.ReturnType).GenerateTypeSyntax();
+                    newReturnType = knownTypes.TaskOfTType.Construct(methodSymbol.ReturnType).GenerateTypeSyntax();
```

Comment:
is all this new code intentionally C# specific?

Reply:
Yes, there's no async-streams in C#

Reply:
why is C# being left out? :(

Reply:
I agree with @CyrusNajmabadi, it would be nice if we had them in C#

Reply:
Hehe, meant VB